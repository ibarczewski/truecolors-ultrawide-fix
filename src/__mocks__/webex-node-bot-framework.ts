export const mockBot = {
  sendCard: jest.fn()
};

export const mockWebex = {
  attachmentActions: {
    get: jest.fn().mockReturnValue(undefined)
  }
};

export const mockFramework = {
  start: jest.fn(),
  getBotByRoomId: jest.fn().mockImplementation(() => mockBot),
  getWebexSDK: jest.fn().mockReturnValue(mockWebex),
  on: jest.fn(),
  hears: jest.fn(),
  debug: jest.fn()
};

const mock = jest.fn().mockImplementation(() => mockFramework);

export default mock;
