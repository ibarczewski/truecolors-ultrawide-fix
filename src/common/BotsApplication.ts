import express from 'express';
import BotApplication from './BotApplication';

export default class BotsApplication {
  public expressApp;
  private server;
  private runningBotApplications: BotApplication[] = [];
  constructor(botApplications: BotApplication[]) {
    this.expressApp = express();
    this.expressApp.use(express.urlencoded({ extended: true }));
    this.expressApp.use(express.json());

    botApplications.forEach((botApplication) => {
      botApplication.start();
      this.expressApp.use('/' + botApplication.slug, botApplication.router);
      this.runningBotApplications.push(botApplication);
    });
  }

  public listen(port) {
    this.expressApp.listen(port, () => {
      this.runningBotApplications.forEach((botApplication) => {
        botApplication.debug('botApplication listening on port %s', port);
      });
    });
  }

  public shutdown() {
    this.runningBotApplications.forEach((botApplication) => {
      botApplication.debug('stoppping...');
    });
    this.server.close();
    Promise.all(
      this.runningBotApplications.map((framework) => framework.stop)
    ).then(() => {
      process.exit();
    });
  }
}
