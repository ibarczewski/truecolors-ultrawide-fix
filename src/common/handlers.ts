export const commonHelpFunction = (welcome) => (framework) => {
  framework.hears('help me', (bot) => {
    bot.say('i ma here to help', welcome);
  });
};
