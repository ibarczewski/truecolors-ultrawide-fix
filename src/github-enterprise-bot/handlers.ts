import { BotHandler } from '../common/BotApplication';

export const handlers: BotHandler[] = [
  {
    command: 'hello',
    handler: (bot) => {
      bot.say('world');
    }
  },
  {
    command: 'get webhook url',
    handler: (bot) => {
      try {
        bot.say(
          `your webhook url is ${process.env.FRAMEWORK_WEBHOOK_URL}/github/${bot.room.id}`
        );
      } catch (error) {
        console.log(error);
      }
    }
  }
];
