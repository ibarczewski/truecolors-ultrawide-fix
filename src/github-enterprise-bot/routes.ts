import { BotRoute } from '../common/BotApplication';
import { githubEnterpriseWebhookController } from './controllers';

export const routes: BotRoute[] = [
  {
    action: 'post',
    path: '/:roomId',
    controller: githubEnterpriseWebhookController
  }
];
