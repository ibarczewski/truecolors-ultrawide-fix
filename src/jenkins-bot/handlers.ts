import { BotHandler } from '../common/BotApplication';
import fetch from 'node-fetch';
import { sendJenkinsConfigurationFormUseCase } from './useCases';

export const handlers: BotHandler[] = [
  {
    command: 'hello',
    handler: (bot) => {
      bot.say('world');
    }
  },
  {
    command: 'demo job completed success',
    handler: (bot) => {
      try {
        // NOTE: using node-fetch here to use our app end to end (rather than simply sending a card directly)
        fetch(`${process.env.FRAMEWORK_WEBHOOK_URL}/jenkins/${bot.room.id}`, {
          method: 'post',
          body: JSON.stringify({
            build: {
              full_url: 'https://google.com/',
              number: 3,
              phase: 'COMPLETED',
              status: 'SUCCESS'
            },
            name: 'fake jenkins job name'
          }),
          headers: { 'Content-Type': 'application/json' }
        });
      } catch (e) {
        console.log(e);
      }
    }
  },
  {
    command: 'demo job completed failure',
    handler: (bot) => {
      try {
        // NOTE: using node-fetch here to use our app end to end (rather than simply sending a card directly)
        fetch(`${process.env.FRAMEWORK_WEBHOOK_URL}/jenkins/${bot.room.id}`, {
          method: 'post',
          body: JSON.stringify({
            build: {
              full_url: 'https://google.com/',
              number: 4,
              phase: 'COMPLETED',
              status: 'FAILURE'
            },
            name: 'fake jenkins job name'
          }),
          headers: { 'Content-Type': 'application/json' }
        });
      } catch (e) {
        console.log(e);
      }
    }
  },
  {
    command: 'demo job queued',
    handler: (bot) => {
      try {
        // NOTE: using node-fetch here to use our app end to end (rather than simply sending a card directly)
        fetch(`${process.env.FRAMEWORK_WEBHOOK_URL}/jenkins/${bot.room.id}`, {
          method: 'post',
          body: JSON.stringify({
            build: {
              full_url: 'https://google.com/',
              number: 4,
              phase: 'QUEUED'
            },
            name: 'fake jenkins job name'
          }),
          headers: { 'Content-Type': 'application/json' }
        });
      } catch (e) {
        console.log(e);
      }
    }
  },
  {
    command: 'get webhook url',
    handler: (bot) => {
      bot.say(`${process.env.FRAMEWORK_WEBHOOK_URL}/jenkins/${bot.room.id}`);
    }
  },
  {
    command: 'configure',
    handler: (bot, trigger) => {
      sendJenkinsConfigurationFormUseCase.execute(trigger.person.id, bot);
    }
  }
];
