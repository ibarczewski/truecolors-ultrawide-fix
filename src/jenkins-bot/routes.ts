import { BotRoute } from '../common/BotApplication';
import { jenkinsNotificationController } from './controllers';

export const routes: BotRoute[] = [
  {
    action: 'post',
    path: '/:roomId',
    controller: jenkinsNotificationController
  },
  {
    action: 'post',
    path: '/:roomId/:settingsEnvelopeID',
    controller: jenkinsNotificationController
  }
];
