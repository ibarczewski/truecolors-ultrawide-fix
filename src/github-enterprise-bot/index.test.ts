import request from 'supertest';
import { handlers } from './handlers';
import { routes } from './routes';
import { mockBot } from 'webex-node-bot-framework';
import BotApplication from '../common/BotApplication';
import BotsApplication from '../common/BotsApplication';

jest.mock('webex-node-bot-framework');

function createMockGithubBotApp() {
  const githubEnterpriseBot = new BotApplication(
    'github',
    'foo',
    null,
    routes,
    handlers
  );
  return new BotsApplication([githubEnterpriseBot]).expressApp;
}

describe('issue assigned notification', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('when issue is assigned, posts an issue assigned notification card', async () => {
    const app = createMockGithubBotApp();
    await request(app)
      .post('/github/foo')
      .send({
        action: 'assigned',
        issue: {
          title: 'fooTitle',
          html_url: 'fooHtml',
          number: 5
        },
        repository: {
          name: 'fooRepoName',
          html_url: 'fooURL'
        },
        assignee: {
          login: 'fooUser',
          html_url: 'fooUserUrl'
        },
        sender: {
          login: 'barUser',
          html_url: 'barUserUrl'
        }
      });

    expect(mockBot.sendCard.mock.calls).toMatchSnapshot();
  });

  test('to be deprecated: when issue is unassigned, does not post issue assigned', async () => {
    const app = createMockGithubBotApp();

    await request(app)
      .post('/github/foo')
      .send({
        action: 'unassigned',
        issue: {
          title: 'fooTitle',
          html_url: 'fooHtml',
          number: 5
        },
        repository: {
          name: 'fooRepoName',
          html_url: 'fooURL'
        },
        assignee: {
          login: 'fooUser',
          html_url: 'fooUserUrl'
        },
        sender: {
          login: 'barUser',
          html_url: 'barUserUrl'
        }
      });

    expect(mockBot.sendCard).not.toHaveBeenCalled();
  });

  test('when pull request opened, posts a pull request notification card', async () => {
    const app = createMockGithubBotApp();
    await request(app)
      .post('/github/foo')
      .send({
        action: 'opened',
        pull_request: {
          title: 'fooTitle',
          html_url: 'fooHtml',
          number: 5
        },
        repository: {
          name: 'fooRepoName',
          html_url: 'fooURL'
        },
        sender: {
          login: 'barUser',
          html_url: 'barUserUrl'
        }
      });

    expect(mockBot.sendCard.mock.calls).toMatchSnapshot();
  });
});
