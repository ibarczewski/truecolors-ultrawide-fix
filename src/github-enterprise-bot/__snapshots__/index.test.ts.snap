// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue assigned notification when issue is assigned, posts an issue assigned notification card 1`] = `
Array [
  Array [
    Object {
      "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
      "body": Array [
        Object {
          "columns": Array [
            Object {
              "items": Array [
                Object {
                  "color": "Accent",
                  "text": "[fooRepoName](fooURL)",
                  "type": "TextBlock",
                  "weight": "Lighter",
                },
                Object {
                  "color": "Light",
                  "size": "Large",
                  "spacing": "Small",
                  "text": "Issue assigned to [fooUser](fooUserUrl)",
                  "type": "TextBlock",
                  "weight": "Bolder",
                  "wrap": true,
                },
              ],
              "type": "Column",
              "width": "stretch",
            },
          ],
          "type": "ColumnSet",
        },
        Object {
          "columns": Array [
            Object {
              "items": Array [
                Object {
                  "color": "Light",
                  "text": "Assigned by:",
                  "type": "TextBlock",
                },
                Object {
                  "color": "Light",
                  "spacing": "Small",
                  "text": "Issue:",
                  "type": "TextBlock",
                },
              ],
              "type": "Column",
              "width": 35,
            },
            Object {
              "items": Array [
                Object {
                  "color": "Light",
                  "text": "[barUser](barUserUrl)",
                  "type": "TextBlock",
                },
                Object {
                  "color": "Light",
                  "spacing": "Small",
                  "text": "[#5 - fooTitle](fooHtml)",
                  "type": "TextBlock",
                },
              ],
              "type": "Column",
              "width": 65,
            },
          ],
          "horizontalAlignment": "Center",
          "spacing": "Padding",
          "type": "ColumnSet",
        },
      ],
      "type": "AdaptiveCard",
      "version": "1.2",
    },
  ],
]
`;

exports[`issue assigned notification when pull request opened, posts a pull request notification card 1`] = `
Array [
  Array [
    Object {
      "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
      "body": Array [
        Object {
          "columns": Array [
            Object {
              "items": Array [
                Object {
                  "color": "Accent",
                  "text": "[fooRepoName](fooURL)",
                  "type": "TextBlock",
                  "weight": "Lighter",
                },
                Object {
                  "color": "Light",
                  "size": "Large",
                  "spacing": "Small",
                  "text": "Pull Request opened: [fooTitle](fooHtml)",
                  "type": "TextBlock",
                  "weight": "Bolder",
                  "wrap": true,
                },
              ],
              "type": "Column",
              "width": "stretch",
            },
          ],
          "type": "ColumnSet",
        },
        Object {
          "columns": Array [
            Object {
              "items": Array [
                Object {
                  "color": "Light",
                  "text": "Opened by:",
                  "type": "TextBlock",
                },
                Object {
                  "color": "Light",
                  "spacing": "Small",
                  "text": "PR#:",
                  "type": "TextBlock",
                },
              ],
              "type": "Column",
              "width": 35,
            },
            Object {
              "items": Array [
                Object {
                  "color": "Light",
                  "text": "[barUser](barUserUrl)",
                  "type": "TextBlock",
                },
                Object {
                  "color": "Light",
                  "spacing": "Small",
                  "text": "5",
                  "type": "TextBlock",
                },
              ],
              "type": "Column",
              "width": 65,
            },
          ],
          "horizontalAlignment": "Center",
          "spacing": "Padding",
          "type": "ColumnSet",
        },
        Object {
          "actions": Array [
            Object {
              "title": "Open in Github Enterprise",
              "type": "Action.OpenUrl",
              "url": "fooHtml",
            },
          ],
          "type": "ActionSet",
        },
      ],
      "type": "AdaptiveCard",
      "version": "1.2",
    },
  ],
]
`;
